mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,661 @@
+            query=processed_query.query, span=span, entity=entity
+        )
+        return query_entity
+
+    def _resolve_conflicts(self, current_entities, annotated_entities):
+        """ Resolve overlaps between existing entities and newly annotad entities.
+
+        Args:
+            current_entities (list): List of existing query entities.
+            annotated_entities (list): List of new query entities.
+
+        Returns:
+            final_entities (list): List of resolved query entities.
+        """
+        overwrite = self.config["overwrite"]
+        base_entities = annotated_entities if overwrite else current_entities
+        other_entities = current_entities if overwrite else annotated_entities
+
+        additional_entities = []
+        for o_entity in other_entities:
+            no_overlaps = [
+                Annotator._no_overlap(o_entity, b_entity) for b_entity in base_entities
+            ]
+            if all(no_overlaps):
+                additional_entities.append(o_entity)
+        return base_entities + additional_entities
+
+    @staticmethod
+    def _no_overlap(entity_one, entity_two):
+        """ Returns True if two query entities do not overlap.
+        """
+        return (
+            entity_one.span.start > entity_two.span.end
+            or entity_two.span.start > entity_one.span.end
+        )
+
+    @staticmethod
+    def _unannotate_query(processed_query, entity_types):
+        """ Removes specified entities in a processed query.
+
+        Args:
+            processed_query (ProcessedQuery): A processed query.
+            entity_types (list): List of entities to remove.
+        """
+        if entity_types == ["*"]:
+            processed_query.entities = ()
+        final_entities = []
+        for query_entity in processed_query.entities:
+            if query_entity.entity.type not in entity_types:
+                final_entities.append(query_entity)
+        processed_query.entities = tuple(final_entities)
+
+    @abstractmethod
+    def parse(self, sentence, **kwargs):
+        """ Extract entities from a sentence. Detected entities should be
+        represented as dictionaries with the following keys: "body", "start"
+        (start index), "end" (end index), "value", "dim" (entity type).
+
+        Args:
+            sentence (str): Sentence to detect entities.
+
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class SpacyAnnotator(Annotator):
+    """ Annotator class that uses spacy to generate annotations.
+    """
+
+    def __init__(self, app_path, config=None, model="en_core_web_lg"):
+        super().__init__(app_path=app_path, config=config)
+
+        self.nlp = SpacyAnnotator._load_model(model)
+        self.model = model
+        self.duckling = DucklingRecognizer.get_instance()
+        self.SYS_MAPPINGS = {
```

Comment:
Maybe `SYS_MAPPINGS ` could be renamed as `ANNOTATOR_TO_DUCKLING_ENTITY_MAPPINGS `